crates/control/src/role_assignment.rs

```diff
@@ -662,114 +687,66 @@ fn pick_role_with_penalties(
         unassigned_robots -= 1;
     }
 
-    if unassigned_robots > 0
-        && role_assignment[PlayerNumber::One].is_none()
-        && penalties[PlayerNumber::One].is_none()
-    {
+    // Keeper
+    if unassigned_robots > 0 && needs_assignment(PlayerNumber::One, penalties, &role_assignment) {
```

Comment:
If you find yourself explaining code structure with comments, think about introducing smaller function scopes by extracting code passages into function calls.

Reply:
done